src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -918,15 +922,18 @@ void GcsActorManager::ReconstructActor(
   }
 }
 
-void GcsActorManager::OnActorCreationFailed(std::shared_ptr<GcsActor> actor,
-                                            bool destroy_actor) {
-  if (destroy_actor) {
-    DestroyActor(actor->GetActorID());
-  } else {
+void GcsActorManager::OnActorSchedulingFailed(std::shared_ptr<GcsActor> actor,
+                                              bool runtime_env_setup_failed) {
+  if (!runtime_env_setup_failed) {
     // We will attempt to schedule this actor once an eligible node is
     // registered.
     pending_actors_.emplace_back(std::move(actor));
+    return;
   }
+
+  // If there is runtime env failure, mark this actor as dead immediately.
+  ReconstructActor(actor->GetActorID(), /*need_reschedule=*/false,
```

Comment:
`DestroyActor` is a public method used to kill actors, it can be called at any time. No matter actor is created or not. It'll do additional cleanups to that actor.

`ReconstructActor (need_reschedule=false)` is a private method, it can only be called after actor is created. Or some data won't be cleanup.